{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ed850710",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:50.550763Z",
     "start_time": "2023-03-28T19:10:49.590742Z"
    }
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import cobra\n",
    "import libsbml\n",
    "from cobra import Model, Reaction, Metabolite, manipulation\n",
    "import pandas as pd\n",
    "import csv\n",
    "import copy\n",
    "from pprint import pprint\n",
    "\n",
    "\n",
    "\n",
    "#Change working dir first, ty ChatGPT, much loves\n",
    "cwd = os.getcwd()\n",
    "# Split the path into a list of directories\n",
    "directories = cwd.split(os.sep)\n",
    "# Remove the last two directories from the list\n",
    "directories = directories[:-2]\n",
    "# Join the directories back into a path\n",
    "new_cwd = os.sep.join(directories)\n",
    "# Change the current working directory to the new path\n",
    "os.chdir(new_cwd)\n",
    "\n",
    "sys.path.append(\"./src\")\n",
    "\n",
    "import model_manipulation  as mm\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "61cc9aa1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:50.554920Z",
     "start_time": "2023-03-28T19:10:50.551949Z"
    }
   },
   "outputs": [],
   "source": [
    "#These functions are used just to fix the ids in the legacy SBML, which wasn't fixed by CobraPy's SBML module.\n",
    "#This would at least allow me to start my pipeline from actual scratch.\n",
    "_renames = (\n",
    "    (\".\", \"_DOT_\"),\n",
    "    (\"(\", \"_LPAREN_\"),\n",
    "    (\")\", \"_RPAREN_\"),\n",
    "    (\"-\", \"_DASH_\"))\n",
    "\n",
    "id_to_fix = \"EX_co2_LPAREN_e_RPAREN__BS\"\n",
    "\n",
    "\n",
    "\n",
    "def escape_str_id_new(id_str):\n",
    "    \"\"\"make a single string id SBML compliant\"\"\"\n",
    "    for c in (\"'\", '\"'):\n",
    "        if id_str.startswith(c) and id_str.endswith(c) \\\n",
    "                and id_str.count(c) == 2:\n",
    "            id_str = id_str.strip(c)\n",
    "    for char, escaped_char in _renames:\n",
    "        id_str = id_str.replace(escaped_char, char)\n",
    "    return id_str\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6ec48e83",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.672269Z",
     "start_time": "2023-03-28T19:10:50.556757Z"
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "%%capture\n",
    "\n",
    "ios2164_orig = cobra.io.read_sbml_model(\"./Supplementary_PrevLit/Lakshmanan et al (2015)/ios2164.xml\");\n",
    "\n",
    "ios2164_orig.solver = 'gurobi'\n",
    "\n",
    "inf = 1e6\n",
    "#Set all non-zero flux bounds to 1e-6\n",
    "\n",
    "mm.set_flux_to_1e6(ios2164_orig)\n",
    "\n",
    "#Fix reaction & metabolite names\n",
    "for rxns in ios2164_orig.reactions:\n",
    "    id_old = rxns.id\n",
    "    rxns.id = escape_str_id_new(id_old)\n",
    "    \n",
    "for mets in ios2164_orig.metabolites:\n",
    "    id_old = mets.id\n",
    "    mets.id = escape_str_id_new(id_old)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "320f1304",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.684684Z",
     "start_time": "2023-03-28T19:10:56.674920Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Change PEPC stoichiometry from consuming CO2 to consuming HCO3\n",
    "co2_c = ios2164_orig.metabolites.get_by_id('co2_c')\n",
    "hco3_c = ios2164_orig.metabolites.get_by_id('hco3_c')\n",
    "h_c = ios2164_orig.metabolites.get_by_id('h_c')\n",
    "h2o_c = ios2164_orig.metabolites.get_by_id('h2o_c')\n",
    "\n",
    "PEPC_old = ios2164_orig.reactions.PPCc\n",
    "PEPC_temp = PEPC_old\n",
    "\n",
    "# # #Removes old co2, h and h2o in favor of hco3\n",
    "PEPC_temp.add_metabolites({hco3_c:-1, co2_c:1, h2o_c:1, h_c:-1})\n",
    "\n",
    "ios2164_orig.reactions.PPCc = PEPC_temp\n",
    "\n",
    "#PEPC now mass-balanced after adjusting coefficients\n",
    "ios2164_orig.reactions.PPCc.check_mass_balance()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ece52963",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.722475Z",
     "start_time": "2023-03-28T19:10:56.686253Z"
    }
   },
   "outputs": [],
   "source": [
    "#let's add a reaction specific for phloem, which includes only amino acids, starch and sucrose\n",
    "#Values are from Poolman et al. (2014) and is rescaled from mol/gcw * time to mmol/gcw * time (similar to ios2164 and Arnold's Atmodel)\n",
    "\n",
    "\n",
    "#Read from csv\n",
    "read_phl = pd.read_csv('./misc/phloem_sap_comp.csv')\n",
    "phl_dict = dict(read_phl.values)\n",
    "\n",
    "\n",
    "model_rxn_list = list()\n",
    "for items in phl_dict:\n",
    "    #Update reaction signs first so it's all negative \n",
    "    phl_dict[items] = phl_dict[items] * -1\n",
    "    \n",
    "    #Retrieve pointer from original model to retrieve metabolite\n",
    "    model_rxn = ios2164_orig.metabolites.get_by_id(items)\n",
    "    model_rxn_list.append(model_rxn)\n",
    "\n",
    "#Update rxns to pointers to orig model\n",
    "for rxns in model_rxn_list:\n",
    "    phl_dict[rxns] = phl_dict.pop(rxns.id)\n",
    "\n",
    "#Generate new phloem reaction\n",
    "phl_rxn = Reaction('DM_Phloem')\n",
    "phl_rxn.name = 'export reaction for phloem sap'\n",
    "phl_rxn.add_metabolites(phl_dict)\n",
    "phl_rxn.notes['SUBSYSTEM']='Demand'\n",
    "\n",
    "\n",
    "\n",
    "ios2164_orig.add_reactions([phl_rxn])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a2ba6acb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.743734Z",
     "start_time": "2023-03-28T19:10:56.724927Z"
    }
   },
   "outputs": [],
   "source": [
    "#Add NGAM reactions based on B&B's measurements\n",
    "#This is a  good way to account for metabolic costs. However, the only limitation would this be would be that the reaction bounds are constant and non-variable\n",
    "\n",
    "ngam_atp_c = Reaction('ngam_atp_c')\n",
    "ngam_atp_s = Reaction('ngam_atp_s')\n",
    "ngam_atp_x = Reaction('ngam_atp_x')\n",
    "ngam_atp_m = Reaction('ngam_atp_m')\n",
    "\n",
    "#Set names\n",
    "ngam_atp_c.name = 'ATP NGAM for Cytosol'\n",
    "ngam_atp_s.name = 'ATP NGAM for Chloroplast (Plastid)'\n",
    "ngam_atp_x.name = 'ATP NGAM for Peroxisome'\n",
    "ngam_atp_m.name = 'ATP NGAM for Mitochondria'\n",
    "\n",
    "#Retrieve metabolites per compartment\n",
    "atp_c = ios2164_orig.metabolites.atp_c\n",
    "adp_c = ios2164_orig.metabolites.adp_c\n",
    "h2o_c = ios2164_orig.metabolites.h2o_c\n",
    "pi_c = ios2164_orig.metabolites.pi_c\n",
    "h_c = ios2164_orig.metabolites.h_c\n",
    "\n",
    "atp_s = ios2164_orig.metabolites.atp_s\n",
    "adp_s = ios2164_orig.metabolites.adp_s\n",
    "h2o_s = ios2164_orig.metabolites.h2o_s\n",
    "pi_s = ios2164_orig.metabolites.pi_s\n",
    "h_s = ios2164_orig.metabolites.h_s\n",
    "\n",
    "atp_x = ios2164_orig.metabolites.atp_x\n",
    "adp_x = ios2164_orig.metabolites.adp_x\n",
    "h2o_x = ios2164_orig.metabolites.h2o_x\n",
    "pi_x = ios2164_orig.metabolites.pi_x\n",
    "h_x =  ios2164_orig.metabolites.h_x\n",
    "\n",
    "atp_m = ios2164_orig.metabolites.atp_m\n",
    "adp_m = ios2164_orig.metabolites.adp_m\n",
    "h2o_m = ios2164_orig.metabolites.h2o_m\n",
    "pi_m = ios2164_orig.metabolites.pi_m\n",
    "h_m = ios2164_orig.metabolites.h_m\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "ngam_atp_c.add_metabolites({atp_c:-1, h2o_c: -1, adp_c:1, pi_c:1, h_c: 1})\n",
    "ngam_atp_s.add_metabolites({atp_s:-1, h2o_s: -1, adp_s:1, pi_s:1, h_s: 1})\n",
    "ngam_atp_x.add_metabolites({atp_x:-1, h2o_x: -1, adp_x:1, pi_x:1, h_x: 1})\n",
    "ngam_atp_m.add_metabolites({atp_m:-1, h2o_m: -1, adp_m:1, pi_m:1, h_m: 1})\n",
    "\n",
    "#Fix bounds of NGAM reaction. Values are from B&B. \n",
    "#However, what I'll implement instead will be Topfer's linear NGAM approach, which scales NGAM with PPFD\n",
    "\n",
    "ngam_atp_c.notes['SUBSYSTEM']='Maintainance reaction'\n",
    "ngam_atp_s.notes['SUBSYSTEM']='Maintainance reaction'\n",
    "ngam_atp_x.notes['SUBSYSTEM']='Maintainance reaction'\n",
    "ngam_atp_m.notes['SUBSYSTEM']='Maintainance reaction'\n",
    "\n",
    "ios2164_orig.add_reactions([ngam_atp_c])\n",
    "ios2164_orig.add_reactions([ngam_atp_s])\n",
    "ios2164_orig.add_reactions([ngam_atp_x])\n",
    "ios2164_orig.add_reactions([ngam_atp_m])\n",
    "\n",
    "#Change bounds of NGAM reaction to one way only.\n",
    "ngam_atp_c.bounds = (0,inf) #0.0427\n",
    "ngam_atp_s.bounds = (0,inf) #0.1527\n",
    "ngam_atp_x.bounds = (0,inf) #0.0076\n",
    "ngam_atp_m.bounds = (0,inf) #0.0091"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9f19b614",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.761383Z",
     "start_time": "2023-03-28T19:10:56.746470Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NGAM NADPHOX mass balance:  {}\n"
     ]
    }
   ],
   "source": [
    "#NGAM for NADPH should be defined in the following compartments:\n",
    "\n",
    "#cytosol, mitochondria and plastid\n",
    "\n",
    "ngam_nadphox_c = Reaction('ngam_nadphox_c')\n",
    "ngam_nadphox_s = Reaction('ngam_nadphox_s')\n",
    "ngam_nadphox_m = Reaction('ngam_nadphox_m')\n",
    "\n",
    "ngam_nadphox_c.name = \"Generic cytosolic NADPH Oxidase for NGAM\"\n",
    "ngam_nadphox_s.name = \"Generic stromal NADPH Oxidase for NGAM\"\n",
    "ngam_nadphox_m.name = \"Generic mitochondrial NADPH Oxidase for NGAM\"\n",
    "\n",
    "#reaction would be the following\n",
    "# #NADPH + 2O2 ↔ NADP+ + 2O2− + H+\n",
    "#From Cheung et al (2013)'s work on modelling the effects of NGAM on plant fluxes, which in turn was  followed \n",
    "# NADP_c + x_NADPHWorkc + WATER_c ⇔ NADPH_c + ×met_2259 + PROTON_c\t\n",
    "#Met_2259 is oxygen\n",
    "\n",
    "#Get all the constituent metabolites for the generic NADPH oxidase reaction in each compartment\n",
    "nadph_c = ios2164_orig.metabolites.nadph_c\n",
    "nadph_s = ios2164_orig.metabolites.nadph_s\n",
    "nadph_m = ios2164_orig.metabolites.nadph_m\n",
    "\n",
    "\n",
    "nadp_c = ios2164_orig.metabolites.nadp_c\n",
    "nadp_s = ios2164_orig.metabolites.nadp_s\n",
    "nadp_m = ios2164_orig.metabolites.nadp_m\n",
    "\n",
    "h_c = ios2164_orig.metabolites.h_c\n",
    "h_s = ios2164_orig.metabolites.h_s\n",
    "h_m = ios2164_orig.metabolites.h_m\n",
    "\n",
    "o2_c = ios2164_orig.metabolites.o2_c\n",
    "o2_s = ios2164_orig.metabolites.o2_s\n",
    "o2_m = ios2164_orig.metabolites.o2_m\n",
    "\n",
    "h2o2_c = ios2164_orig.metabolites.h2o2_c\n",
    "h2o2_s =  ios2164_orig.metabolites.h2o2_s\n",
    "h2o2_m =  ios2164_orig.metabolites.h2o2_m\n",
    "\n",
    "\n",
    "ngam_nadphox_c.add_metabolites({nadph_c:-1, o2_c:-1, h_c:-1, nadp_c:1, h2o2_c: 1})\n",
    "ngam_nadphox_s.add_metabolites({nadph_s:-1, o2_s:-1, h_s:-1, nadp_s:1, h2o2_s: 1})\n",
    "ngam_nadphox_m.add_metabolites({nadph_m:-1, o2_m:-1, h_m:-1, nadp_m:1, h2o2_m: 1})\n",
    "\n",
    "ngam_nadphox_c.bounds=(0, inf)\n",
    "ngam_nadphox_s.bounds=(0,inf)\n",
    "ngam_nadphox_m.bounds=(0,inf)\n",
    "\n",
    "ngam_nadphox_c.notes['SUBSYSTEM']='Maintainance reaction'\n",
    "ngam_nadphox_s.notes['SUBSYSTEM']='Maintainance reaction'\n",
    "ngam_nadphox_m.notes['SUBSYSTEM']='Maintainance reaction'\n",
    "\n",
    "ios2164_orig.add_reactions([ngam_nadphox_c])\n",
    "ios2164_orig.add_reactions([ngam_nadphox_s])\n",
    "ios2164_orig.add_reactions([ngam_nadphox_m])\n",
    "\n",
    "#Notes: Nov 22, 2022 -- Changed NADPHOX to produce H2O2 instead of H2O in order for it to become\n",
    "#Balanced.\n",
    "print('NGAM NADPHOX mass balance: ', ngam_nadphox_c.check_mass_balance())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "256fe7c6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.766393Z",
     "start_time": "2023-03-28T19:10:56.762979Z"
    }
   },
   "outputs": [],
   "source": [
    "#Add constraints such that the NADPH oxidase flux ratios are equally distributed in each compartment\n",
    "#Methodology is the same as the one by Cheung et al (2014) along with other studies stemming\n",
    "#from it (Shameer et al 2018, Topfer et al 2020)\n",
    "\n",
    "nadphox_c_s = mm.set_fix_flux_ratio({ngam_nadphox_c.id:1, ngam_nadphox_s.id:1},ios2164_orig)\n",
    "nadphox_c_s.name = \"nadphox_cs_ratio\"\n",
    "nadphox_s_m = mm.set_fix_flux_ratio({ngam_nadphox_s.id:1, ngam_nadphox_m.id:1}, ios2164_orig)\n",
    "nadphox_s_m.name = \"nadphox_sm_ratio\"\n",
    "\n",
    "ios2164_orig.add_cons_vars(nadphox_c_s)\n",
    "ios2164_orig.add_cons_vars(nadphox_s_m)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e4cf457a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.774297Z",
     "start_time": "2023-03-28T19:10:56.767735Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/latex": [
       "$\\displaystyle - 1.0 \\cdot ngam_{nadphox c} + 1.0 \\cdot ngam_{nadphox c reverse 4016e} + 1.0 \\cdot ngam_{nadphox s} - 1.0 \\cdot ngam_{nadphox s reverse 1b6e8}$"
      ],
      "text/plain": [
       "-1.0*ngam_nadphox_c + 1.0*ngam_nadphox_c_reverse_4016e + 1.0*ngam_nadphox_s - 1.0*ngam_nadphox_s_reverse_1b6e8"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Does not work\n",
    "#Add constraint such that ATP:NADPH oxidase ratio is at 3:1\n",
    "\n",
    "fex_nadphoxc =  mm.get_flux_exp(ios2164_orig, ngam_nadphox_c)\n",
    "fex_nadphoxs =  mm.get_flux_exp(ios2164_orig, ngam_nadphox_s)\n",
    "fex_nadphoxm =  mm.get_flux_exp(ios2164_orig, ngam_nadphox_m)\n",
    "fex_atpc = mm.get_flux_exp(ios2164_orig, ngam_atp_c)\n",
    "fex_atps =  mm.get_flux_exp(ios2164_orig, ngam_atp_s)\n",
    "fex_atpm = mm.get_flux_exp(ios2164_orig, ngam_atp_m)\n",
    "fex_atpx =mm.get_flux_exp(ios2164_orig, ngam_atp_x)\n",
    "\n",
    "\n",
    "nadphox_atpase = ios2164_orig.problem.Constraint(3*(fex_nadphoxc+fex_nadphoxs +fex_nadphoxm) \n",
    "                                     - 1*(fex_atpc + fex_atps + fex_atpm),\n",
    "                                     lb=0,ub=0)\n",
    "nadphox_atpase.name = \"nadphox_atpase_ratio\"\n",
    "\n",
    "\n",
    "ios2164_orig.add_cons_vars(nadphox_atpase)\n",
    "ios2164_orig.constraints.nadphox_cs_ratio.expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9b49f5a7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.779063Z",
     "start_time": "2023-03-28T19:10:56.775099Z"
    }
   },
   "outputs": [],
   "source": [
    "#Restrict all reactions to white light only\n",
    "for lights in ios2164_orig.reactions:\n",
    "    if 'PRISM' in lights.id:\n",
    "        ios2164_orig.reactions.get_by_id(lights.id).bounds = (-inf,inf)\n",
    "        #Constrain all non-white light reactions to 0\n",
    "        if 'white' not in lights.id:\n",
    "            ios2164_orig.reactions.get_by_id(lights.id).bounds = (0,0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "43a6faf1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.790326Z",
     "start_time": "2023-03-28T19:10:56.780218Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>RBPOs</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Ribulose-biphosphate Oxygenase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x7ff52ecd8880</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>o2_s + rb15bp_s --> 2pglyc_s + 3pg_s + 2 h_s</p>\n",
       "                    <p style='text-align:right'>O2 + D-Ribulose 1,5-bisphosphate --> 2-Phosphoglycolate + 3-Phospho-D-glycerate + 2 H+</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>(LOC_Os01g58020 and LOC_Os05g35330 and LOC_Os11g32770 and LOC_Os12g10580 and LOC_Os10g21280) and...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction RBPOs at 0x7ff52ecd8880>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Add Rubisco Oxygenase-specific reaction to model\n",
    "\n",
    "#First retrieve the carboxylase reaction\n",
    "rbc_rxn = ios2164_orig.reactions.RBPCs\n",
    "\n",
    "\n",
    "#Define reaction. Id is RBPOs\n",
    "rbo_rxn = Reaction('RBPOs')\n",
    "rbo_rxn.name = 'Ribulose-biphosphate Oxygenase'\n",
    "\n",
    "#Components of oxygenase reaction are as follows\n",
    "#o2[s] + h2o[s] + rb15bp[s] -> 3pg[s] + 2pglyc[s]\n",
    "#Interestingly, Arnold's model accounts for the generation of protons by the rubisco reaction.\n",
    "#There's also support from literature on how protonation may also enhance CO2 capture by a plant Cell.\n",
    "#To accomodate this I'll just add the reaction specifically for both RBO and RBC reactions.\n",
    "\n",
    "#Retrieve metabolites\n",
    "o2_s = ios2164_orig.metabolites.o2_s\n",
    "h2o_s = ios2164_orig.metabolites.h2o_s\n",
    "rb15bp_s = ios2164_orig.metabolites.rb15bp_s\n",
    "h_s = ios2164_orig.metabolites.h_s\n",
    "_3pg_s = ios2164_orig.metabolites.get_by_id('3pg_s')\n",
    "_2pglyc_s = ios2164_orig.metabolites.get_by_id('2pglyc_s')\n",
    "\n",
    "#Define Stoichiometry\n",
    "rbo_rxn.add_metabolites({o2_s:-1.0, rb15bp_s:-1, h_s:2, _3pg_s: 1, _2pglyc_s: 1})\n",
    "\n",
    "#Copy other details from rbc_rxn\n",
    "rbo_rxn.subsystem = rbc_rxn.subsystem\n",
    "rbo_rxn.bounds = rbc_rxn.bounds\n",
    "#Define encoded genes; copied from ios2164 supplementary\n",
    "rbo_rxn.gene_reaction_rule = rbc_rxn.gene_name_reaction_rule\n",
    "rbo_rxn.notes['SUBSYSTEM'] = 'Calvin cycle'\n",
    "#Add to model\n",
    "ios2164_orig.add_reactions([rbo_rxn])\n",
    "\n",
    "ios2164_orig.repair\n",
    "\n",
    "ios2164_orig.reactions.RBPOs\n",
    "        \n",
    "#Balanced equation is as follows:\n",
    "#o2_s + rb15bp_s --> 2pglyc_s + 3pg_s + 2 h_s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d96713e7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.834572Z",
     "start_time": "2023-03-28T19:10:56.791183Z"
    }
   },
   "outputs": [],
   "source": [
    "def rebalance_model(model):\n",
    "    '''\n",
    "    \n",
    "    This function is used to rebalance any reactions\n",
    "    found in the old model to reduce any possible futile cycling. \n",
    "    \n",
    "    This curation step is done with the aid of Metacyc and BRENDA\n",
    "    as well as the check_mass_balance() function from CobraPy\n",
    "    to validate any inconsistent chemical equations as well as\n",
    "    compound formulas that weren't encoded properly. \n",
    "    \n",
    "    The only remaining reactions that weren't curated were as  follows:\n",
    "    - reactions with undefined side groups encoded as \"R\" and were therefore expectedly\n",
    "    unbalanced since CobraPy does not detect them\n",
    "    - Charge imbalances that were validated to be balanced based on the databases above as well as\n",
    "     w/ other extant curated models\n",
    "    - Spectral decomposition reactions w/c were expected to drain out of the system w/ excess photon input\n",
    "    - Other reactions that are not expected to carry flux with the model constraints.\n",
    "\n",
    "    Overall this rebalancing step reduces the amount of unbalanced internal reactions from 28% to ~13% and removes.\n",
    "    - Removes any excess fluxes towards proton generating reactions when using pFBA\n",
    "    - Prevents spurious Biomass generation with O2 input\n",
    "    \n",
    "    \n",
    "    '''\n",
    "    ios2164_orig.metabolites.get_by_id('4c2me_s').formula = 'C14H23N3O14P2'\n",
    "    ios2164_orig.metabolites.udprmn_c.formula = 'C15H22N2O16P2' #Works\n",
    "    ios2164_orig.metabolites.cgly_c.formula = 'C5H10N2O3S' #Not encoded properly\n",
    "    ios2164_orig.metabolites.cholp_r.formula = 'C5H13NO4P'\n",
    "    ios2164_orig.metabolites.cholp_s.formula = 'C5H13NO4P'\n",
    "    ios2164_orig.metabolites.dtmp_s.formula = 'C10H13N2O8P' #Was not encoded properly in model\n",
    "    ios2164_orig.metabolites.get_by_id('5dpmev_c').formula = 'C6H10O10P2'\n",
    "    ios2164_orig.metabolites.lnlncoa_r.formula = 'C39H62N7O17P3S'\n",
    "    ios2164_orig.metabolites.lnlnlACP_s.formula = 'C29H52N2O8PRS'\n",
    "    ios2164_orig.metabolites.lnlnl_c.formula = 'C18H29O2'\n",
    "    ios2164_orig.metabolites.lnlnl_s.formula = 'C18H29O2'\n",
    "    ios2164_orig.metabolites.lnlnl_x.formula = 'C18H29O2'\n",
    "    ios2164_orig.metabolites.o2_u.formula = 'O2' #Metabolite Not encoded properly in model\n",
    "    ios2164_orig.metabolites.plast_s.formula = 'C53H82O2'\n",
    "    \n",
    "    \n",
    "    ios2164_orig.reactions.AASADc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.AATCm.add_metabolites({'h_m':-1})\n",
    "    ios2164_orig.reactions.AATCs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.ACDOc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.ACGAMKc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.ACHBSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ACHYSLc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.ACRSPc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.ACSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ADEDAc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.ADMATs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ADPARTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ALATm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.ALATs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ALTNDAc.add_metabolites({'h_c':-2.0})\n",
    "    ios2164_orig.reactions.AMETAHYCSTc.add_metabolites({'h_c':1,  'h2o_c':1})\n",
    "    ios2164_orig.reactions.ANTPPTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ANXANDEOs.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.AOBUTDx.add_metabolites({'h_x':1})\n",
    "    ios2164_orig.reactions.APPRTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ARGSSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ASCOXc.add_metabolites({'h_c':8})\n",
    "    ios2164_orig.reactions.ASCPO1c.add_metabolites({'h_c':4})\n",
    "    ios2164_orig.reactions.ASCPO1s.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.ASCPO2c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.ASCPO2s.add_metabolites({'h_s':3})\n",
    "    ios2164_orig.reactions.ASNS1s.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ASNTLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ASNtm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.ASPALs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ASPCLc.add_metabolites({'h_c': -1})\n",
    "    ios2164_orig.reactions.ASPCT1c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.ASPCTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ASPHGNAHr.add_metabolites({'h_r':-1})\n",
    "    ios2164_orig.reactions.ASPHGNAr.add_metabolites({'h_r':-1})\n",
    "    ios2164_orig.reactions.ASPTLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.ASPTLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ATAHc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.ATPAPc.reaction = 'atp_c + 2.0 h2o_c --> amp_c + 2.0 pi_c + 2.0 h_c'\n",
    "    ios2164_orig.reactions.ATPSs.add_metabolites({'h_u':-1})\n",
    "    ios2164_orig.reactions.BRS2c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.C3STDH2c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.C4STMO1c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CAFFMTc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CAVNS2s.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.CBMKs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.CBPSs.add_metabolites({'h_s':3})\n",
    "    ios2164_orig.reactions.CC3H1c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.CDPDAGSm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.CELLSc.add_metabolites({'h_c': 2})\n",
    "    ios2164_orig.reactions.CHALIc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.CHOLPCTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.CKDH1c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKDH2c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKDH3c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKGT1c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKGT2c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKGT3c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKGT4c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKGT5c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKGT6c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKGT7c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKGT8c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKOGT1c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKOGT2c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKOGT3c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CKPc.add_metabolites({'h_c':2})\n",
    "    ios2164_orig.reactions.CMPAc.add_metabolites({'h_c':-1})\n",
    "#     ios2164_orig.reactions.CNDNG3Gc.add_metabolites({'h_c': 2})\n",
    "    ios2164_orig.reactions.CNDNG5Gc.add_metabolites({'h_c':2})\n",
    "    ios2164_orig.reactions.CNDNGLT5c.add_metabolites({'h_c':2})\n",
    "    ios2164_orig.reactions.COUMGUTFc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.COUMNSc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.CTPCHc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.CYSSs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.CYSTLm.add_metabolites({'h_m':-3})\n",
    "    ios2164_orig.reactions.CYTTHs.bounds = (-inf, 0)\n",
    "    ios2164_orig.reactions.DAGLc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.DCTPDAc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.DGSNTTPc.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.DHDHc.add_metabolites({'h_c':2})\n",
    "    ios2164_orig.reactions.DHDPSs.add_metabolites({'h_s': -2})\n",
    "    ios2164_orig.reactions.DHKFORc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.DHKRc.add_metabolites({'h_c': 1})\n",
    "    ios2164_orig.reactions.DHNCOAHs.add_metabolites({'h_s': 1})\n",
    "    ios2164_orig.reactions.DHNCOASs.add_metabolites({'h_s': -1})\n",
    "    ios2164_orig.reactions.DHNOPTs.add_metabolites({'h_s': 2})\n",
    "    ios2164_orig.reactions.DHNPPTs.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.DHORSts.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.DHPDc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.DHPRs.add_metabolites({'h_s': 2})\n",
    "    ios2164_orig.reactions.DHPRys.add_metabolites({'h_s': 2})\n",
    "    ios2164_orig.reactions.DHPSm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.DHQORc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.DHQRc.add_metabolites({'h2o_c':-1})\n",
    "    ios2164_orig.reactions.DMATTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.DMATTm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.DHPAHs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.DMPPTs.add_metabolites({'h_s': 1})\n",
    "    ios2164_orig.reactions.DMPQMTs.add_metabolites({'h_s': 1})\n",
    "    ios2164_orig.reactions.DNTPPAc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.EAR160ym.add_metabolites({'h_m':-1})\n",
    "    ios2164_orig.reactions.EAR160ys.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.ECASDATEO1c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.ECASDATEO3c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.ECASDO1c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.ECASDO3c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.ECASDSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.EIKAUROc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.EIKAURSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.EKAURSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ENTBSs.delete #Remove bacterial reaction w/ no GPR\n",
    "    ios2164_orig.reactions.ESANDO1c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.ESANDSs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.ESANDSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.ETHAPTr.add_metabolites({'h_r':1})\n",
    "    ios2164_orig.reactions.FA140COAabcx.add_metabolites({'h2o_x':-1,'h_x':1})\n",
    "    ios2164_orig.reactions.FA160ACOASc.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.FA160ACPHs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.FA160COAabcx.add_metabolites({'h2o_x':-1,'h_x':1})\n",
    "    ios2164_orig.reactions.FA161COAabcx.add_metabolites({'h2o_x':-1,'h_x':1})\n",
    "    ios2164_orig.reactions.FA180COAabcx.add_metabolites({'h2o_x':-1,'h_x':1})\n",
    "    ios2164_orig.reactions.FA181COAabcx.add_metabolites({'h2o_x':-1,'h_x':1})\n",
    "    ios2164_orig.reactions.FA182ACOASc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.FA182ACPHs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.FA182COAabcx.add_metabolites({'h2o_x':-1,'h_x':1})\n",
    "    ios2164_orig.reactions.FA183ACOASc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.FA183ACPHs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.FA183COAabcx.add_metabolites({'h2o_x':-1,'h_x':1})\n",
    "    ios2164_orig.reactions.FA200COAHr.add_metabolites({'h_r':1})\n",
    "    ios2164_orig.reactions.FA220COAHr.add_metabolites({'h_r':1})\n",
    "    ios2164_orig.reactions.FA240COAHr.add_metabolites({'h_r':1})\n",
    "    ios2164_orig.reactions.FA260COAHr.add_metabolites({'h_r':1})\n",
    "    ios2164_orig.reactions.FA280COAHr.add_metabolites({'h_r':1})\n",
    "    ios2164_orig.reactions.FA300COAHr.add_metabolites({'h_r':1})\n",
    "    ios2164_orig.reactions.FAS161COA_Lc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.FAS181ACP_Lm.add_metabolites({'h_m':-1})\n",
    "    ios2164_orig.reactions.FAS181ACP_Ls.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.FAS181COA_Lc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.FAS182ACP_Ls.add_metabolites({'h_s':-2})\n",
    "    ios2164_orig.reactions.FAS182COA_Lc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.FAS183ACP_Ls.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.FAS183COA_Lc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.FERHc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.FGFTm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.FGFTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.FMNATc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.FMNATm.add_metabolites({'h_m':2})\n",
    "    ios2164_orig.reactions.FMNATs.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.FNORs.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.FORMHLc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.FRDPFTs.add_metabolites({'h_s': 5})\n",
    "    ios2164_orig.reactions.FRPTc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.FRTTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.FTHFCLc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.FTHFCLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.FTHFCLs.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.GA4DOc.add_metabolites({'h2o_c':1, 'h_c':1})\n",
    "    ios2164_orig.reactions.GAAKORG10c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.GAAKORG11c.add_metabolites({'h2o_c':1})\n",
    "    ios2164_orig.reactions.GAAKORG12c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.GAAKORG8c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.GAAKORG9c.add_metabolites({'h2o_c':1})\n",
    "    ios2164_orig.reactions.GAMPTc.add_metabolites({'h_c':1.0})\n",
    "    ios2164_orig.reactions.GBDPs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.GCOADCx.add_metabolites({'h_x':-1})\n",
    "    ios2164_orig.reactions.GDPMDHc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.GENT2c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.GENT3c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.GLCNSc.add_metabolites({'h_c':2})\n",
    "    ios2164_orig.reactions.GLUTRSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.GLYATx.add_metabolites({'h_x':-1})\n",
    "    ios2164_orig.reactions.GLYB1s.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.GLYDHDm.add_metabolites({'h_m':-1})\n",
    "    ios2164_orig.reactions.GLYTLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.GLYTLs.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.GMPc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.GNSKc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.GRTTm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.GRTTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.GSNPc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.GSNRHc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.GTHDHc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.GTHDHs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.GTPCYTDPTc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.GTPCYTDPTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.GTPDHs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.GTPDPKs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.H2Sts.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.HCYSMT2c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.HDHs.add_metabolites({'h_s': -1})\n",
    "    ios2164_orig.reactions.HEMEOSm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.HFERMTc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.HISTLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.HISTLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.HMGCOARc.add_metabolites({'h_c': -1.0})\n",
    "    ios2164_orig.reactions.HMGCOARyc.add_metabolites({'h_c': -1.0})\n",
    "    ios2164_orig.reactions.HMGLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.HOXGs.add_metabolites({'h_s':-2})\n",
    "    ios2164_orig.reactions.HPPKm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.IAAAMYHc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.IAAIASc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.IAAIGSc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.ILETLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.ILETLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.IPDPT1s.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.IPDPT2s.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.IPKK2c.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.IPKK4c.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.IPKK5c.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.IPKK6c.add_metabolites({'h_c': -2.0})\n",
    "    ios2164_orig.reactions.ITK1Kc.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.LCYCORc.add_metabolites({'h_c': 1})\n",
    "    ios2164_orig.reactions.LCYSTSm.add_metabolites({'h_m':2})\n",
    "    ios2164_orig.reactions.LCYSTSs.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.LIMSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.LNL9HPLs.add_metabolites({'h_s':-2})\n",
    "    ios2164_orig.reactions.LPLGDOc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.LYSTLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.LYSTLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.MACPCLm.add_metabolites({'h_m':-2})\n",
    "    ios2164_orig.reactions.MACPCLs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.MACPMMEc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.MAGLc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.MANKINc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.MARMSc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.MCOATAc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.MCOATAm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.MCOATAs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.MDHARm.add_metabolites({'h_m':-2})\n",
    "    ios2164_orig.reactions.MDHARc.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.MDHARyc.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.MDHARym.add_metabolites({'h_m':-2})\n",
    "    ios2164_orig.reactions.MECDPDH5s.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.MEPCTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.METSc.add_metabolites({'h_c':1.0})\n",
    "    ios2164_orig.reactions.METSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.METTLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.METTLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.MGCHLs.add_metabolites({'h_s': 2})\n",
    "    ios2164_orig.reactions.MGCm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.MI1345PP2c.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.MI1345PPc.add_metabolites({'h_c': -2.0})\n",
    "    ios2164_orig.reactions.MI1346PK1c.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.MI1346PK2c.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.MI1456PKc.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.MI34Pc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.MLTHFR_Fc.add_metabolites({'h_c':3})\n",
    "    ios2164_orig.reactions.MLTHFRc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.MLTHFRs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.MOMASs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.MOMASys.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.MPMECY1s.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.MPMECY2s.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.MPMTs.add_metabolites({'h_s': -1})\n",
    "    ios2164_orig.reactions.MRs.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.MSBQMTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.MTAMnh4m.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.NADOR2u.add_metabolites({'h_u':-1.0})\n",
    "    ios2164_orig.reactions.NNAMc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.NMNDAc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.NNDPRs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.NO3Rm.add_metabolites({'h_m':-2})\n",
    "    ios2164_orig.reactions.NOSc.reaction = '2 arg-L_c + 3 nadph_c + 4 o2_c + 1 h_c --> 2 citr-L_c + 3 nadp_c + 2 no_c + 4 h2o_c'\n",
    "    ios2164_orig.reactions.NRGNAORc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.NRGNMTc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.NRGNNORc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.OCTDPOTs.add_metabolites({'h_s': 1})\n",
    "    ios2164_orig.reactions.ORPRTs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.OXAACc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.P450MO1c.add_metabolites({'h_c':- 1})\n",
    "    ios2164_orig.reactions.P450MO2c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.P450MO3c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.P450MO4c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.P450MO5c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.P450MO6c.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.PCAHr.add_metabolites({'h_r':-1})\n",
    "    ios2164_orig.reactions.PEDCm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.PEDCr.add_metabolites({'h_r':1})\n",
    "    ios2164_orig.reactions.PEPSs.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.PHENH4Lc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.PHETLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.PHETLs.add_metabolites({'h_s':1}) \n",
    "    ios2164_orig.reactions.PHYTS1s.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.PHYTS2s.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.PINCLc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.PLIBTNSm.add_metabolites({'h_m':-3})\n",
    "    ios2164_orig.reactions.PLIVLPNSm.add_metabolites({'h_m':-3})\n",
    "    ios2164_orig.reactions.PPASm.add_metabolites({'h_m':-1})\n",
    "    ios2164_orig.reactions.PPASs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.PPDKc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.PPDKs.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.PPPGPPPc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.PRADPs.add_metabolites({'h_s': 1})\n",
    "    ios2164_orig.reactions.PRAICs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.PRDPARs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.PRFGSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.PRIMADNSs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.PRIMARTSc.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.PROTLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.PROTLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.QRCTNGLT3c.add_metabolites({'h_c': 1})\n",
    "    ios2164_orig.reactions.RBPCs.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.RIBFKc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.RIBFSs.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.SELNPS.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.SERTLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.SERTLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.SHCHFs.add_metabolites({'h_s':3})\n",
    "    ios2164_orig.reactions.SOXs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.SQTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.STEMOc.add_metabolites({'h_c': -1})\n",
    "    ios2164_orig.reactions.STEMSs.add_metabolites({'h_s': 1})\n",
    "    ios2164_orig.reactions.SUATs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.SULRs.add_metabolites({'h_s':6})\n",
    "    ios2164_orig.reactions.TAGLc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.THR3DHx.add_metabolites({'h_x':-1.0})\n",
    "    ios2164_orig.reactions.THRTLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.TMDPKc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.TMPPPs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.TRICINS1c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.TRICINS2c.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.TRPTLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.TRPTLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.TYRLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.TYRNH4Lc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.TYRTLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.TYRTLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.UAMASc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.UDPGALMc.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.UDPGALts.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.UDPGATc.add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.UDPGts.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.UDPSQSs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.UPRTs.add_metabolites({'h_s':-1})\n",
    "    ios2164_orig.reactions.URDGLAHc.add_metabolites({'h_c':-1})     \n",
    "    ios2164_orig.reactions.UREASEc.add_metabolites({'h_c':-2.0})\n",
    "    ios2164_orig.reactions.UROc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.VALTLm.add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.VALTLs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.VLXANDEOs.add_metabolites({'h_s':2})\n",
    "    ios2164_orig.reactions.XSc.add_metabolites({'h_c':2})\n",
    "    ios2164_orig.reactions.get_by_id('2HEDBT8GTc').add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.get_by_id('2HNDBT8GTc').add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.get_by_id('2HNDGDHc').add_metabolites({'h_c':1})\n",
    "    ios2164_orig.reactions.get_by_id('2SBZCOALs').add_metabolites({'h_s': 1})\n",
    "    ios2164_orig.reactions.get_by_id('3OAS180m').add_metabolites({'h_m':1})\n",
    "    ios2164_orig.reactions.get_by_id('3OAS180s').add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.get_by_id('SUCL(ADP)m').add_metabolites({'h_m':-1})\n",
    "    ios2164_orig.reactions.get_by_id('SUCL(GDP)m').add_metabolites({'h_m':-1})\n",
    "    \n",
    "    \n",
    "    #Last iteration\n",
    "    ios2164_orig.reactions.ACSERLc.add_metabolites({'h_c':-1})\n",
    "    ios2164_orig.reactions.PRIMARDSs.add_metabolites({'h_s': 1})\n",
    "    ios2164_orig.reactions.DLDLBTs.add_metabolites({'h_s': -1})\n",
    "    ios2164_orig.reactions.PC2Ds.add_metabolites({'h_s': 2})\n",
    "    ios2164_orig.reactions.GSNDc.add_metabolites({'h_c':-2})\n",
    "    ios2164_orig.reactions.CYSTLs.add_metabolites({'h_c': 1})\n",
    "    ios2164_orig.reactions.FRDPFTm.add_metabolites({'h_m':5})\n",
    "    ios2164_orig.reactions.ARGTs.add_metabolites({'h_s':1})\n",
    "    ios2164_orig.reactions.AMCPCOc.add_metabolites({'h_c':2})\n",
    "    \n",
    "\n",
    "#     ios2164_orig.metabolites.cdpdag_os_r.formula = 'C14H18N3O15P2R2'\n",
    "#     ios2164_orig.metabolites.ocdcya_s.formula = 'C18H29O2' #Doesn't work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ddaea047",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.885374Z",
     "start_time": "2023-03-28T19:10:56.835376Z"
    }
   },
   "outputs": [],
   "source": [
    "#Let's add a specific demand reactions for excess oxygen and protons\n",
    "\n",
    "dm_o2 = Reaction('DM_o2(c)')\n",
    "dm_o2.name = \"Demand reaction for oxygen\"\n",
    "\n",
    "o2_c = ios2164_orig.metabolites.o2_c\n",
    "dm_o2.add_metabolites({o2_c:-1.0})\n",
    "dm_o2.bounds = (0, inf)\n",
    "dm_o2.notes['SUBSYSTEM'] = 'Demand'\n",
    "ios2164_orig.add_reactions([dm_o2])\n",
    "\n",
    "#Demand reactions for protons. If needed, I can add this to the model. This was from B&B(2019) and was\n",
    "#incorporated to deal with proton recycling in the model.\n",
    "\n",
    "dm_h_c = Reaction('DM_h(c)')\n",
    "dm_h_c.name = \"Demand reaction for excess protons (cytosolic)\"\n",
    "h_c = ios2164_orig.metabolites.h_c\n",
    "dm_h_c.add_metabolites({h_c:-1})\n",
    "dm_h_c.bounds = (0, 0)\n",
    "ios2164_orig.add_reactions([dm_h_c])\n",
    "\n",
    "dm_h_s = Reaction('DM_h(s)')\n",
    "dm_h_s.name = \"Demand reaction for excess protons (stromal)\"\n",
    "h_s = ios2164_orig.metabolites.h_s\n",
    "dm_h_s.add_metabolites({h_s:-1})\n",
    "dm_h_s.bounds = (0, 0)\n",
    "ios2164_orig.add_reactions([dm_h_s])\n",
    "\n",
    "dm_h_x = Reaction('DM_h(x)')\n",
    "dm_h_x.name = \"Demand reaction for excess protons (Peroxisomal)\"\n",
    "h_x = ios2164_orig.metabolites.h_x\n",
    "dm_h_x.add_metabolites({h_x:-1})\n",
    "dm_h_x.bounds = (0, 0)\n",
    "ios2164_orig.add_reactions([dm_h_x])\n",
    "\n",
    "dm_h_m = Reaction('DM_h(m)')\n",
    "dm_h_m.name = \"Demand reaction for excess protons (mitochodrial)\"\n",
    "h_m = ios2164_orig.metabolites.h_m\n",
    "dm_h_m.add_metabolites({h_m:-1})\n",
    "dm_h_m.bounds = (0, 0)\n",
    "ios2164_orig.add_reactions([dm_h_m])\n",
    "\n",
    "\n",
    "dm_h_u= Reaction('DM_h(u)')\n",
    "dm_h_u.name = \"Demand reaction for excess protons (thylakoid)\"\n",
    "h_u = ios2164_orig.metabolites.h_u\n",
    "dm_h_u.add_metabolites({h_u:-1})\n",
    "dm_h_u.bounds = (0, 0)\n",
    "ios2164_orig.add_reactions([dm_h_u])\n",
    "\n",
    "dm_h_v= Reaction('DM_h(v)')\n",
    "dm_h_v.name = \"Demand reaction for excess protons (vacuole)\"\n",
    "h_v = ios2164_orig.metabolites.h_v\n",
    "dm_h_v.add_metabolites({h_v:-1})\n",
    "dm_h_v.bounds = (0, 0)\n",
    "ios2164_orig.add_reactions([dm_h_v])\n",
    "\n",
    "dm_h_r= Reaction('DM_h(r)')\n",
    "dm_h_r.name = \"Demand reaction for excess protons (ER)\"\n",
    "h_r = ios2164_orig.metabolites.h_r\n",
    "dm_h_r.add_metabolites({h_r:-1})\n",
    "dm_h_r.bounds = (0, 0)\n",
    "ios2164_orig.add_reactions([dm_h_r])\n",
    "\n",
    "\n",
    "dm_h2o_c = Reaction('DM_h2o(c)')\n",
    "dm_h2o_c.name = \"Demand reaction for excess H2O production\"\n",
    "h2o_c = ios2164_orig.metabolites.h2o_c\n",
    "dm_h2o_c.add_metabolites({h2o_c:-1})\n",
    "dm_h2o_c.bounds = (-inf,inf)\n",
    "ios2164_orig.add_reactions([dm_h2o_c])\n",
    "\n",
    "ios2164_orig.repair()\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "de769dd8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:10:56.982756Z",
     "start_time": "2023-03-28T19:10:56.886321Z"
    }
   },
   "outputs": [],
   "source": [
    "# #Turn off all EC -> cytosol reactions that are used by Sink tissues as these may lead to \n",
    "# #Erroneous cellular efflux of photoassimilates.\n",
    "# #Maybe I can just set it as unidirectional kaya? Either way this would prevent assimilation rin\n",
    "\n",
    "\n",
    "tex_reactions = [rxns for rxns in ios2164_orig.reactions if \"tex\" in rxns.id]\n",
    "tex_reactions\n",
    "\n",
    "sink_reactions_to_turn_off = ['SUCRtex', 'FRUtex', 'GLCtex', 'ASNtex', 'GLNtex', 'ETOHtex',\n",
    "                              'ASNtex', 'ACtex', 'LACtex', 'TSULtex', 'ASNtex', 'LACtex',\n",
    "                              'ALAtex', 'ARGtex', 'ASPtex', 'CYStex', 'GLUtex', 'HIStex',\n",
    "                              'ILEtex', 'LEUtex', 'LYStex', 'METtex', 'PHEtex', 'PROtex', \n",
    "                              'SERtex', 'THRtex',  'TRPtex', 'TYRtex', 'VALtex']\n",
    "\n",
    "\n",
    "for rxns in sink_reactions_to_turn_off:\n",
    "    ios2164_orig.reactions.get_by_id(rxns).bounds = (0,inf)\n",
    "\n",
    "\n",
    "#Based on Yang et al (2020), Phloem loading may occur apoplastically so this assumption may not make too much sense.\n",
    "#Yang, G., Wei, Q., Huang, H., & Xia, J. (2020). Amino acid transporters in plant cells: A brief review. Plants, 9(8), 967.\n",
    "#Upon reviewing this, I think the paper is only pertaining to the apoplastic transport at the level of the \n",
    "#Phloem cell\n",
    "\n",
    "#Rebalance all reactions encoded in the above step\n",
    "\n",
    "rebalance_model(ios2164_orig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "353d15bd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:11:17.057818Z",
     "start_time": "2023-03-28T19:10:56.983649Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /tmp/tmpeujwjxio.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 2002 rows, 4924 columns, 21402 nonzeros\n",
      "Read LP format model from file /tmp/tmpojd93to4.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 2002 rows, 4924 columns, 21402 nonzeros\n",
      "Read LP format model from file /tmp/tmpjr7msp6_.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 2002 rows, 4924 columns, 21402 nonzeros\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'EX_co2(e)' since it already exists.\n",
      "Ignoring reaction 'EX_h2o(e)' since it already exists.\n",
      "Ignoring reaction 'EX_h(e)' since it already exists.\n",
      "Ignoring reaction 'EX_no3(e)' since it already exists.\n",
      "Ignoring reaction 'EX_o2(e)' since it already exists.\n",
      "Ignoring reaction 'EX_pi(e)' since it already exists.\n",
      "Ignoring reaction 'EX_sucr(e)' since it already exists.\n",
      "Ignoring reaction 'EX_fru-B(e)' since it already exists.\n",
      "Ignoring reaction 'EX_glc-A(e)' since it already exists.\n",
      "Ignoring reaction 'EX_so4(e)' since it already exists.\n",
      "Ignoring reaction 'EX_hco3(e)' since it already exists.\n",
      "Ignoring reaction 'EX_so3(e)' since it already exists.\n",
      "Ignoring reaction 'EX_h2s(e)' since it already exists.\n",
      "Ignoring reaction 'EX_nh4(e)' since it already exists.\n",
      "Ignoring reaction 'EX_asn-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_gln-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_etoh(e)' since it already exists.\n",
      "Ignoring reaction 'EX_ac(e)' since it already exists.\n",
      "Ignoring reaction 'EX_lac-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_tsul(e)' since it already exists.\n",
      "Ignoring reaction 'EX_fe2(e)' since it already exists.\n",
      "Ignoring reaction 'EX_fe3(e)' since it already exists.\n",
      "Ignoring reaction 'EX_mg2(e)' since it already exists.\n",
      "Ignoring reaction 'EX_ala-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_arg-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_asp-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_cys-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_glu-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_gly(e)' since it already exists.\n",
      "Ignoring reaction 'EX_his-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_ile-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_leu-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_lys-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_met-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_phe-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_pro-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_ser-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_thr-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_trp-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_tyr-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_val-L(e)' since it already exists.\n",
      "Ignoring reaction 'EX_photonVis(e)' since it already exists.\n"
     ]
    }
   ],
   "source": [
    "ios2164_m = copy.deepcopy(ios2164_orig)\n",
    "for met in ios2164_m.metabolites:\n",
    "    if met.compartment != \"e\": #Extracellular, defines boundary reactions\n",
    "        met.id = str(met.id) + '0'\n",
    "        met.compartment = str(met.compartment) + '0'\n",
    "\n",
    "    \n",
    "        \n",
    "for rxn in ios2164_m.reactions:\n",
    "    if \"EX_\" not in rxn.id: #Exclude exchange reactions\n",
    "        rxn.id = str(rxn.id)  + '_M'\n",
    "\n",
    "\n",
    "#Change name of reaction \n",
    "ios2164_m.repair()\n",
    "ios2164_bs = copy.deepcopy(ios2164_orig)\n",
    "\n",
    "#Rename BS reactions and metabolites\n",
    "for met in ios2164_bs.metabolites:\n",
    "    if met.compartment != \"e\": #Extracellular, defines boundary reactions\n",
    "        met.id = str(met.id) + '1'\n",
    "        met.compartment = str(met.compartment) + '1'\n",
    "    \n",
    "for rxn in ios2164_bs.reactions:\n",
    "    if \"EX_\" not in rxn.id: #Exclude exchange reactions\n",
    "        rxn.id = str(rxn.id)  + '_BS'\n",
    "        \n",
    "ios2164_bs.repair()\n",
    "\n",
    "ios2164_2cell = ios2164_m.merge(ios2164_bs, inplace=True, objective = 'sum')\n",
    "\n",
    "#Apparently compartments don't update after  merging so I've updated it na lang after that\n",
    "\n",
    "ios2164_2cell.compartments = {\n",
    " 'c0': 'M Cell Cytoplasm',\n",
    " 's0': 'M Cell Plastid',\n",
    " 'r0': 'M Cell Endoplasmic_reticulum',\n",
    " 'm0': 'M Cell Mitochondrion',\n",
    " 'e': 'Extracellular',\n",
    " 'u0': 'M Cell Thylakoid',\n",
    " 'x0': 'M Cell Peroxisome',\n",
    " 'v0': 'M Cell Vacuole',\n",
    " 'c1': 'BS Cell Cytoplasm',\n",
    " 's1': 'BS Cell Plastid',\n",
    " 'r1': 'BS Cell Endoplasmic_reticulum',\n",
    " 'm1': 'BS Cell Mitochondrion',\n",
    " 'u1': 'BS Cell Thylakoid',\n",
    " 'x1': 'BS Cell Peroxisome',\n",
    " 'v1': 'BS Cell Vacuole'}\n",
    "\n",
    "\n",
    "#Also need to rebuild the model medium\n",
    "ios2164_2cell.repair()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5bc2390",
   "metadata": {},
   "source": [
    "From Blatke & Brautigam (2019):\n",
    "\n",
    "\"The mesophyll and bundle sheath networks are connected by a range of cytosolic transport metabolites including amino acids, sugars (glucose, fructose, sucrose, trehalose, ribose), single phosphorylated sugar (glucose-6-phosphate, glucose-1-phosphate, fructose-6-phosphate, sucrose-6-phosphate), mono-/di-/tri-carboxylic acids (phosphoenolpyruvate, pyruvate, citrate, cis-aconitate, isocitrate, α-ketoglutarate, succinate, fumarate, malate), glyceric acids (2-Phosphoglycerate, 3-Phosphoglycerate), glycolate, glycerate, glyceraldehyde-3-phosphate, di-hydroxyacetone-phosphate and CO2. \n",
    "\n",
    "Nucleotides, NAD/NADH, NADP/NADPH, pyrophosphate, inorganic phosphate are not considered as transport metabolites. \n",
    "\n",
    "Oxaloacetate has been excluded as transport metabolite since concentrations of oxaloacetate are very low in vivo and it is reasonably unstable in aqueous solutions. \n",
    "\n",
    "Other small molecules that can be imported by the bundle sheath from the environment, as well as protons and HCO3-, are not exchanged between the two cell types.\n",
    "\n",
    "#So bale yung mga organic na soluble lang with reasonably long cellular half-lives."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5a11c6e5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:11:17.269080Z",
     "start_time": "2023-03-28T19:11:17.058817Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "added 73 PD reactions to model\n",
      "ngam_atp_c (0, 0)\n",
      "ngam_atp_s (0, 0)\n",
      "ngam_atp_x (0, 0)\n",
      "ngam_atp_m (0, 0)\n"
     ]
    }
   ],
   "source": [
    "#Open the initial curated list to get the metabolite ids\n",
    "transport_rxns = pd.read_csv('./model/initial_list_transport_rxns.csv')\n",
    "#Ok it gets the reactions now\n",
    "\n",
    "for item in ios2164_orig.metabolites:\n",
    "    if item.compartment == \"c\":\n",
    "        if item.id in transport_rxns.id.values:\n",
    "            pd_rxn = Reaction()\n",
    "            #Generate reaction name\n",
    "            pd_rxn.id = str(item.id[:-2] + '_pd') #Replaces suffix with '_pd'\n",
    "            pd_rxn.name = str('Plasmodesmatal transport of ' + item.name)\n",
    "            pd_rxn.subsystem = str('PD Transport Reaction')\n",
    "            #Reaction is reversible\n",
    "            pd_rxn.lower_bound = -inf\n",
    "            pd_rxn.upper_bound = inf\n",
    "            pd_rxn.notes['SUBSYSTEM']='Plasmodesmatal Transport'\n",
    "            \n",
    "            #Generate metabolite names\n",
    "            \n",
    "            \n",
    "            met_mcell = str(item.id) + '0'\n",
    "            met_bcell = str(item.id) + '1'\n",
    "            \n",
    "            met_mcell = ios2164_2cell.metabolites.get_by_id(met_mcell)\n",
    "            met_bcell = ios2164_2cell.metabolites.get_by_id(met_bcell)\n",
    "            pd_rxn.add_metabolites({met_mcell:-1.0, met_bcell:1.0})\n",
    "            \n",
    "            \n",
    "            ios2164_2cell.add_reactions([pd_rxn])\n",
    "            \n",
    "ios2164_2cell.repair()\n",
    "\n",
    "\n",
    "#Print how many reactions added to model\n",
    "printout = 'added ' + str(len(transport_rxns)) + ' PD reactions to model'\n",
    "print(printout)\n",
    "\n",
    "\n",
    "#Set Sucrose PD to be non-reversible (M -> BS) so as to properly model bulk flow\n",
    "ios2164_2cell.reactions.get_by_id('sucr_pd').bounds = (0, inf)\n",
    "\n",
    "#Set Gas influx is such a way that it allows one-way gas evolution from BS to M cells (from respiration)\n",
    "#But should be restricted to prevent free CO2/O2 from flowing into BS cell. Without this constraint pFBA will\n",
    "#Predict that the most efficient way for Rubisco fixation is to directly import CO2 from the M cell to BS cell symplastically,\n",
    "#Given a BS cell with an unconstrained rubisco species. This constraint is based on the assumption that carbon fixation (either to 3PG or to other C3/C4 acids) takes place \n",
    "#in photosynthetically active tissue and that lateral diffusion of CO2 is negligible (Morison et  al., 2005)\n",
    "ios2164_2cell.reactions.get_by_id('co2_pd').bounds = (-inf, 0)\n",
    "ios2164_2cell.reactions.get_by_id('o2_pd').bounds = (-inf, 0)\n",
    "\n",
    "#Turn off all NGAMs in the meantime. I can return it when I perform the  simulations directly\n",
    "for rxns in ios2164_orig.reactions: \n",
    "    if 'ngam_atp' in rxns.id:\n",
    "        ios2164_orig.reactions.get_by_id(rxns.id).bounds = (0,0)\n",
    "        print(rxns.id, rxns.bounds)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1f22883b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:11:17.298165Z",
     "start_time": "2023-03-28T19:11:17.269970Z"
    }
   },
   "outputs": [],
   "source": [
    "#Set all external fluxes to 0 muna to make it all easy.\n",
    "for rxns in ios2164_2cell.exchanges:\n",
    "    rxn_id = rxns.id\n",
    "    ios2164_2cell.reactions.get_by_id(rxn_id).bounds = (0, 0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a8de4a44",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:11:17.320279Z",
     "start_time": "2023-03-28T19:11:17.299061Z"
    }
   },
   "outputs": [],
   "source": [
    "#Rebalance BS Phloem demand reaction to account for transitory starch accumulation (based on assumed 10:1 ratio between BS and M cell volumes)\n",
    "#Starch production in M cell = 1.863 flux units Starch/unit Phloem\n",
    "#Starch production in BS cell = 0.207 flux units Starch/unit plhoem\n",
    "\n",
    "ios2164_2cell.reactions.DM_Phloem_BS.reaction = '''0.0183 ala-L_c1 + 0.0958 arg-L_c1 + 0.172 asn-L_c1 + 0.121 asp-L_c1 + 0.189 gln-L_c1 \n",
    "+ 0.156 glu-L_c1 + 0.00969 his-L_c1 + 0.0204 ile-L_c1 + 0.0172 leu-L_c1 + 0.0516 lys-L_c1 + 0.0043 phe-L_c1 + 0.0129 pro-L_c1 + 0.116 ser-L_c1 \n",
    "+ 1.863 starch_s0 + 0.207 starch_s1 + 0.618 sucr_c1 + 0.044 thr-L_c1 + 0.00539 tyr-L_c1 + 0.0432 val-L_c1 -->'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b668c5ed",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:11:17.609325Z",
     "start_time": "2023-03-28T19:11:17.321199Z"
    }
   },
   "outputs": [],
   "source": [
    "mets = open('./misc/metabolites.csv', 'w')\n",
    "rxns = open('./misc/reactions.csv', 'w')\n",
    "cons = open('./misc/constraints.csv', 'w')\n",
    "subsystems = open('./misc/subsystems.csv', 'w')\n",
    "\n",
    "met_write = csv.writer(mets)\n",
    "rxns_write = csv.writer(rxns)\n",
    "cons_write = csv.writer(cons)\n",
    "subsystems_write = csv.writer(subsystems)\n",
    "\n",
    "for item in ios2164_2cell.metabolites:\n",
    "    row = [item.id, item.name]\n",
    "    met_write.writerow(row)\n",
    "    \n",
    "\n",
    "for item in ios2164_2cell.reactions:\n",
    "    row = [item.id, item.name, item.build_reaction_string(),item.bounds]\n",
    "    rxns_write.writerow(row)\n",
    "\n",
    "for item  in ios2164_2cell.constraints:\n",
    "    row =  [str(item.name), str(item.expression)]\n",
    "    cons_write.writerow(row)\n",
    "    \n",
    "\n",
    "for subsystems in ios2164_2cell.groups:\n",
    "    row = [subsystems]\n",
    "    subsystems_write.writerow(row)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fce02aee",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:11:21.265183Z",
     "start_time": "2023-03-28T19:11:21.136444Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "imbalanced reactions length for ios2164_orig after rebalancing: 290\n",
      "percent imbalanced reactions 12.59774109470026\n"
     ]
    }
   ],
   "source": [
    "#Checking lang\n",
    "imbalanced_rxns = list()\n",
    "\n",
    "non_exchange_rxns = set(ios2164_orig.reactions) - set(ios2164_orig.boundary) - set(ios2164_orig.demands)\n",
    "#Check mass balance of original ios2164 model\n",
    "\n",
    "for rxns in non_exchange_rxns:\n",
    "        #Retrieve mass balance dictionary\n",
    "        mb_dict = rxns.check_mass_balance()\n",
    "        #Check if imbalanced \n",
    "        if len(mb_dict) != 0:\n",
    "            imbalanced_rxns.append(rxns)\n",
    "\n",
    "\n",
    "#149 imbalanced rxns that are in boundary\n",
    "#636 reactions that are within the cell. Let's check which of these are imbaalanced.\n",
    "print('imbalanced reactions length for ios2164_orig after rebalancing:', len(imbalanced_rxns))\n",
    "print('percent imbalanced reactions', (len(imbalanced_rxns)/len(non_exchange_rxns)) * 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b10293c8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-28T19:11:21.135359Z",
     "start_time": "2023-03-28T19:11:17.610087Z"
    }
   },
   "outputs": [],
   "source": [
    "cobra.io.write_sbml_model(ios2164_2cell, filename= \"./model/ios2164_2cell.xml\")\n",
    "#Save Json for mapping\n",
    "cobra.io.save_json_model(ios2164_2cell, filename=\"./model/ios2164_2cell.json\")\n",
    "cobra.io.save_json_model(ios2164_bs, filename=\"./model/ios2164_bscell.json\")\n",
    "cobra.io.save_json_model(ios2164_m, filename=\"./model/ios2164_mcell.json\")                    \n",
    "#Save one-cell model to incorporate updates\n",
    "cobra.io.write_sbml_model(ios2164_orig, filename=\"./model/ios2164_1cell.xml\")\n",
    "\n",
    "\n",
    "#Add trans reactions then afterwards save the model to json\n",
    "import model_initialize as m_i\n",
    "m_i.add_trans_reactions(ios2164_2cell)\n",
    "cobra.io.save_json_model(ios2164_2cell, filename= \"./model/ios2164_2cell_w-trans.json\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
